### PR TITLE
Optimize

### PR DESCRIPTION
**Pull request by Spottedleaf**
Adds a distance map where players are tracked to coordinates in which
they are close to. The class contains a validatePlayer() method
which was used for testing.

The class uses pooled hash tables to reduce object churning, and was
tested via the following:

https://gist.github.com/Spottedleaf/a93bb7a8993d6ce142d3efc5932bf573

With the distance map, looking up players in a specific location will be
approximately constant time, which makes this patch scale much better with
more players, at the cost of some performance at lower playercounts.

### CODE REVIEW

File: Spigot-Server-Patches/0408-implement-optional-per-player-mob-spawns.patch [hunk 1]
```diff
@@ -84,46 +648,46 @@ index a7fc34f85..612b9b7e3 100644
...
++        int chunkX = (int)Math.floor(entity.locX) >> 4;
++        int chunkZ = (int)Math.floor(entity.locZ) >> 4;
++        int index = entity.getEntityType().getEnumCreatureType().ordinal();
++
++        for (EntityPlayer player : this.playerMobDistanceMap.getPlayersInRange(chunkX, chunkZ)) {
++            ++player.mobCounts[index];
++        }
++    }
++
++    public int getMobCountNear(EntityPlayer entityPlayer, EnumCreatureType enumCreatureType) {
```
Comment (kickash32) [inline Spigot-Server-Patches/0408-implement-optional-per-player-mob-spawns.patch:684]: `updatePlayerMobTypeMap` and `getMobCountNear` are almost carbon copies of similar methods in `WorldServer` and additionally never used. These should either be removed or the ones in `WorldServer` should be removed and these used instead.

File: Spigot-Server-Patches/0408-implement-optional-per-player-mob-spawns.patch [hunk 2]
```diff
@@ -18,61 +40,603 @@ index 318a470ee..ff1e27e7d 100644
...
 --- a/src/main/java/net/minecraft/server/EntityPlayer.java
 +++ b/src/main/java/net/minecraft/server/EntityPlayer.java
-@@ -80,6 +80,7 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
+@@ -80,6 +80,11 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
      public boolean queueHealthUpdatePacket = false;
      public net.minecraft.server.PacketPlayOutUpdateHealth queuedHealthUpdatePacket;
      // Paper end
-+    public int[] mobCounts = new int[EnumCreatureType.values().length]; // Paper
++    // Paper start - mob spawning rework
++    public static final int ENUMCREATURETYPE_TOTAL_ENUMS = EnumCreatureType.values().length;
```
Comment (kickash32) [inline Spigot-Server-Patches/0408-implement-optional-per-player-mob-spawns.patch:615]: Perhaps `ENUMCREATURETYPE_TOTAL_ENUMS` can be moved to `EnumCreatureType.java`?

**Review by kickash32 — CHANGES_REQUESTED** [review]

File: Spigot-Server-Patches/0408-implement-optional-per-player-mob-spawns.patch (hunk 2 above)
Comment (Spottedleaf) [reply Spigot-Server-Patches/0408-implement-optional-per-player-mob-spawns.patch:615]: not importing the class for that

File: Spigot-Server-Patches/0408-implement-optional-per-player-mob-spawns.patch (hunk 1 above)
Comment (Spottedleaf) [reply Spigot-Server-Patches/0408-implement-optional-per-player-mob-spawns.patch:684]: oops left them in

File: Spigot-Server-Patches/0408-implement-optional-per-player-mob-spawns.patch (hunk 2 above)
Comment (kickash32) [reply Spigot-Server-Patches/0408-implement-optional-per-player-mob-spawns.patch:615]: fair enough

File: Spigot-Server-Patches/0408-implement-optional-per-player-mob-spawns.patch (hunk 1 above)
Comment (Spottedleaf) [reply Spigot-Server-Patches/0408-implement-optional-per-player-mob-spawns.patch:684]: resolved

**Review by kickash32 — APPROVED** [review]

**Pull request merged**